File: forge/src/main/java/cn/breadnicecat/candycraftce/datagen/forge/providers/recipe_builders/SugarFurnaceRecipeBuilder.java
package cn.breadnicecat.candycraftce.datagen.forge.providers.recipe_builders;

import cn.breadnicecat.candycraftce.recipe.CRecipeTypes;
import cn.breadnicecat.candycraftce.recipe.RecipeSerializerExt;
import cn.breadnicecat.candycraftce.recipe.recipes.SugarFurnaceRecipe;
import com.google.gson.JsonObject;
import net.minecraft.data.recipes.FinishedRecipe;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.crafting.Ingredient;
import net.minecraft.world.level.ItemLike;
import org.jetbrains.annotations.NotNull;

import java.util.function.Consumer;

/**
 * Created in 2024/2/3
 * Project: candycraftce
 *
 * @author <a href="https://github.com/Bread-NiceCat">Bread_NiceCat</a>
 * <p>
 */
public class SugarFurnaceRecipeBuilder extends CRecipeBuilderBase {
	private Item result;
	private Ingredient ingredient;
	private int xp;

	private SugarFurnaceRecipeBuilder() {
	}

	public static SugarFurnaceRecipeBuilder builder() {
		return new SugarFurnaceRecipeBuilder();
	}

	public SugarFurnaceRecipeBuilder result(ItemLike result) {
		this.result = result.asItem();
		return this;
	}

	public SugarFurnaceRecipeBuilder ingredient(Ingredient ingredient) {
		this.ingredient = ingredient;
		return this;
	}

	public SugarFurnaceRecipeBuilder xp(int exp) {
		this.xp = exp;
		return this;
	}

	@Override
	public @NotNull Item getResult() {
		return result;
	}

	@Override
	public void save(@NotNull Consumer<FinishedRecipe> consumer, @NotNull ResourceLocation recipeId) {
		consumer.accept(new Result(recipeId));
	}

	private class Result extends CFinishedRecipeBase {


		public Result(ResourceLocation id) {
			super(id);
		}

		@Override
		public void serializeRecipeData(@NotNull JsonObject json) {
			getType().toJson(json, new SugarFurnaceRecipe(id, result, ingredient, xp));
		}

		@Override
		public @NotNull RecipeSerializerExt<SugarFurnaceRecipe> getType() {
			return CRecipeTypes.SUGAR_FURNACE_TYPE.getSerializer();
		}
	}
}

File: src/main/kotlin/cn/breadnicecat/candycraftce/data/recipes/SugarFurnaceRecipeBuilder.kt
package cn.breadnicecat.candycraftce.data.recipes

import cn.breadnicecat.candycraftce.core.recipe.recipes.SugarFurnaceRecipe
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.crafting.Ingredient

/**
 * Created by NiceCat on 2026/3/23.
 * Project: candycraftce
 * @author <a href="https://github.com/BreadNiceCat">Bread_NiceCat</a>
 *
 */
class SugarFurnaceRecipeBuilder {
    companion object {
        inline fun builder(action: SugarFurnaceRecipeBuilder.() -> Unit): SugarFurnaceRecipeBuilder {
            return SugarFurnaceRecipeBuilder().apply(action)
        }

        fun direct(
            id: ResourceLocation,
            ingredient: Ingredient,
            result: ItemStack,
            exp: Float = 0f,
        ): SugarFurnaceRecipeBuilder {
            return builder {
                this.id = id
                this.ingredient = ingredient
                this.result = result
                this.exp = exp
            }
        }
    }

    lateinit var id: ResourceLocation
    lateinit var ingredient: Ingredient
    lateinit var result: ItemStack
    var exp: Float = 0f

    fun build(): Result {
        return Result(id, SugarFurnaceRecipe(id, ingredient, result, exp))
    }

    class Result(
        id: ResourceLocation,
        recipe: SugarFurnaceRecipe,
    ) : CodecFinishedRecipe<SugarFurnaceRecipe>(id, recipe) {
        override val codec = SugarFurnaceRecipe.CODEC
        override fun getType() = SugarFurnaceRecipe.Serializer
    }
}